### PR TITLE
LimitToLast: back port some changes from android

### PR DESCRIPTION
**Pull request by wu-hui**
* s/memorized/memoized
* Disallow double-unlisten (remote store) and double-releasing (local store)
* Consolidate duplicated integration tests
* Mark synthesizeTargetChange as non-multitab

### CODE REVIEW

File: packages/firestore/src/local/local_store.ts [hunk 1]
```diff
@@ -917,15 +915,15 @@ export class LocalStore {
           return PersistencePromise.forEach(removed, (key: DocumentKey) =>
             this.persistence.referenceDelegate.removeReference(txn, key)
           ).next(() => {
-            this.persistence.referenceDelegate.removeTarget(txn, targetData);
+            this.persistence.referenceDelegate.removeTarget(txn, targetData!);
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts]: Can you check if this is still needed? With TypeScript 3.7, `assert` should now be type-aware.

File: packages/firestore/src/local/local_store.ts [hunk 2]
```diff
@@ -917,15 +915,15 @@ export class LocalStore {
...
           });
         } else {
           return PersistencePromise.resolve();
         }
       })
       .then(() => {
         this.targetDataByTarget = this.targetDataByTarget.remove(targetId);
-        this.targetIdByTarget.delete(targetData.target);
+        this.targetIdByTarget.delete(targetData!.target);
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store.ts]: Same as above.

File: packages/firestore/src/remote/remote_event.ts [hunk 3]
```diff
@@ -127,11 +127,12 @@ export class TargetChange {
 
   /**
    * HACK: Views require TargetChanges in order to determine whether the view is
-   * CURRENT, but secondary tabs don't receive remote events. So this method is
-   * used to create a synthesized TargetChanges that can be used to apply a
-   * CURRENT status change to a View, for queries executed in a different tab.
+   * CURRENT, but secondary tabs or new queries mapped to an active target don't
+   * receive remote events (yet). So this method is used to create a synthesized
+   * TargetChanges that can be used to apply a CURRENT status change to a View,
+   * for queries executed in a different tab, or for the new queries to raise
```
Comment (schmidt-sebastian) [inline packages/firestore/src/remote/remote_event.ts]: How about:

This method is used to create a synthesized TargetChanges that can be used to apply a CURRENT status change to a View (for queries executed in a different tab) or for new queries (to raise snapshots with correct CURRENT status).

File: packages/firestore/src/remote/remote_event.ts [hunk 4]
```diff
@@ -127,11 +127,12 @@ export class TargetChange {
 
   /**
    * HACK: Views require TargetChanges in order to determine whether the view is
-   * CURRENT, but secondary tabs don't receive remote events. So this method is
-   * used to create a synthesized TargetChanges that can be used to apply a
-   * CURRENT status change to a View, for queries executed in a different tab.
+   * CURRENT, but secondary tabs or new queries mapped to an active target don't
```
Comment (schmidt-sebastian) [inline packages/firestore/src/remote/remote_event.ts]: Nit: Would be easier to parse if you made it "to an already active target".

File: packages/firestore/src/local/local_store.ts (hunk 1 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts]: Leads to compilation error.

File: packages/firestore/src/local/local_store.ts (hunk 2 above)
Comment (wu-hui) [reply packages/firestore/src/local/local_store.ts]: Same.

File: packages/firestore/src/remote/remote_event.ts (hunk 3 above)
Comment (wu-hui) [reply packages/firestore/src/remote/remote_event.ts]: Done.

File: packages/firestore/src/remote/remote_event.ts (hunk 4 above)
Comment (wu-hui) [reply packages/firestore/src/remote/remote_event.ts]: Took your comment below.

**Pull request merged**